### PR TITLE
Replace MongoDB/WCF reflection lookups with DuckTyping

### PR DESCRIPTION
**Pull request by zacharycmontoya**
This allows us to completely delete all of the types within the Datadog.Trace.ClrProfiler.Emit namespace 🎉

### CODE REVIEW

File: tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Wcf/SyncMethodInvokerIntegration.cs
Comment (pierotibou) [inline tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Wcf/SyncMethodInvokerIntegration.cs:61]: [nit] This whole body could be moved in WcfCommon and used here and in AsyncMethodInvoker_InvokeBegin_Integration.cs and TaskMethodInvokerIntegration.cs (even the TODOs are duplicated :))

**Review by pierotibou — COMMENTED** [review]
Looks great, thanks.
Just added a minor nit, not even related to your change per se :p. 
Waiting for the tests to pass for approval.

**Review by tonyredondo — APPROVED** [review]
Thanks for this! LGTM

**Review by lucaspimentel — APPROVED** [review]
Nice, thanks!

File: tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Wcf/SyncMethodInvokerIntegration.cs
Comment (lucaspimentel) [reply tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Wcf/SyncMethodInvokerIntegration.cs:61]: Good catch! Maybe in a separate PR since it's not related to removing `Datadog.Trace.ClrProfiler.Emit`?

File: tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/MongoDb/IBsonDocument.cs
Comment (andrewlock) [inline tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/MongoDb/IBsonDocument.cs:15]: We've tended towards interfaces in the past, but @tonyredondo has some benchmarks that show that really we should be using a standard `class` with `virtual` methods instead (where we can't use copy structs), as they are quite a bit faster.

File: tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/MongoDb/IBsonElement.cs
Comment (andrewlock) [inline tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/MongoDb/IBsonElement.cs:15]: I think this could be a `[DuckCopy]` struct?

File: tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/MongoDb/IConnection.cs
Comment (andrewlock) [inline tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/MongoDb/IConnection.cs:16]: I think this could be a `[DuckCopy]` struct, though then we can't use it as a constraint in the integrations, so would need to do manual duck typing in `MongoDbIntegration.CreateScope()`. 

Pros of struct: Would only do the duck type if the integration is enabled/not already in mongo context, less overhead
Cons: More faff, not sure if the type constraint approach would have any performance benefits @tonyredondo, or if it's just convenience?

**Review by andrewlock — APPROVED** [review]
LGTM, a couple of questions around struct/interface duck typing, but they're more general questions about best practices we should take across all integrations!

File: tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/MongoDb/IConnection.cs
Comment (tonyredondo) [reply tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/MongoDb/IConnection.cs:16]: The constraints approach have similar performance than the DuckCopy but allowing methods calls. This because internally the generic T is an struct type. I prefer to use the constraint approach if we can, the limitation is that we cannot compare with null unless we also apply the IDuckType interface as a constraint.

File: tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/MongoDb/DatabaseNamespaceStruct.cs
Comment (tonyredondo) [inline tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/MongoDb/DatabaseNamespaceStruct.cs:17]: We can change the visibility to internal now :)

File: tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/MongoDb/IConnection.cs
Comment (zacharycmontoya) [reply tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/MongoDb/IConnection.cs:16]: @tonyredondo can you clarify, what is your recommendation here? Should I keep as is or make this a struct?

File: tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/MongoDb/DatabaseNamespaceStruct.cs
Comment (zacharycmontoya) [reply tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/MongoDb/DatabaseNamespaceStruct.cs]: Good call, done in 4220883d48a8ea9a1235d12e118a53679f64ad31

File: tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/MongoDb/IBsonElement.cs
Comment (zacharycmontoya) [reply tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/MongoDb/IBsonElement.cs]: I think you're right. Done in 46794d772bf3ede266cdd4202f691014f1a5eb75

File: tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/MongoDb/IBsonDocument.cs
Comment (zacharycmontoya) [reply tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/MongoDb/IBsonDocument.cs]: Done in 46794d772bf3ede266cdd4202f691014f1a5eb75

**Comment by andrewlock** [comment]
## Code Coverage Report :bar_chart:

:heavy_check_mark: Merging #2183 into [master](https://github.com/DataDog/dd-trace-dotnet/tree/b17a9fd4f332efd23af6b4974350f86a54297a22) will will **increase** line coverage by `1%`
:heavy_check_mark: Merging #2183 into [master](https://github.com/DataDog/dd-trace-dotnet/tree/b17a9fd4f332efd23af6b4974350f86a54297a22) will will **increase** branch coverage by `1%`
:heavy_check_mark: Merging #2183 into [master](https://github.com/DataDog/dd-trace-dotnet/tree/b17a9fd4f332efd23af6b4974350f86a54297a22) will will **decrease** complexity by `348`


|           | [master](https://github.com/DataDog/dd-trace-dotnet/tree/b17a9fd4f332efd23af6b4974350f86a54297a22) | #2183       | Change   | 
|:----------|:-----------:|:-----------:|:--------:|
| Lines     | `9334` / `12808` | `8980` / `12218` |          |
| Lines %   | `73%`      | `73%`      |  `1%` :heavy_check_mark:  |
| Branches  | `4246` / `6260` | `4083` / `5923` |          |
| Branches %| `68%`      | `69%`      |  `1%` :heavy_check_mark:  |
| Complexity|   `7456`      | `7108`        |  `-348`  :heavy_check_mark:    |

View the full report for further details:

* [HTML report for master](https://dev.azure.com/datadoghq/dd-trace-dotnet/_build/results?buildId=84518&view=codecoverage-tab) | [Source file](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/84518/artifacts?artifactName=Code%20Coverage%20Report_84518&api-version=6.0&%24format=zip)  
* [HTML report for this PR #2183](https://dev.azure.com/datadoghq/dd-trace-dotnet/_build/results?buildId=84562&view=codecoverage-tab) | [Source file](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/84562/artifacts?artifactName=Code%20Coverage%20Report_84562&api-version=6.0&%24format=zip)

### Datadog.Trace Breakdown :heavy_check_mark:

|        | [master](https://github.com/DataDog/dd-trace-dotnet/tree/b17a9fd4f332efd23af6b4974350f86a54297a22) | #2183       | Change   | 
|:-------|:-----------:|:-----------:|:--------:|
| Lines %| `73%`      | `73%`       |  `1%` :heavy_check_mark:  |
| Branches %| `68%`      | `69%`       |  `1%` :heavy_check_mark:  |
| Complexity| `7456`      | `7108`       |  `-348` :heavy_check_mark:  |

The following classes have significant coverage changes.

| File    | Line coverage change | Branch coverage change | Complexity change |
|:--------|:--------------------:|:----------:|:--------:|
| [Datadog.Trace.ClrProfiler.AutoInstrumentation.Couchbase.IIOServiceExecuteIntegrationBis](https://github.com/DataDog/dd-trace-dotnet/tree/d6d45a5d3fc9e8c60ae9b1be3e25dc403acf3290/tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Couchbase/IIOServiceExecuteIntegrationBis.cs) | `-100%` :no_entry: | `0%` :heavy_check_mark:   | `0` :heavy_check_mark: |
| [Datadog.Trace.ClrProfiler.AutoInstrumentation.Testing.MsTestV2.TestMethodRunnerExecuteIntegration](https://github.com/DataDog/dd-trace-dotnet/tree/d6d45a5d3fc9e8c60ae9b1be3e25dc403acf3290/tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Testing/MsTestV2/TestMethodRunnerExecuteIntegration.cs) | `-33%` :no_entry: | `-19%` :no_entry:   | `0` :heavy_check_mark: |
| [Datadog.Trace.ClrProfiler.AutoInstrumentation.Wcf.SyncMethodInvokerIntegration](https://github.com/DataDog/dd-trace-dotnet/tree/d6d45a5d3fc9e8c60ae9b1be3e25dc403acf3290/tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Wcf/SyncMethodInvokerIntegration.cs) | `-12%` :no_entry: | `-8%` :no_entry:   | `2` :no_entry: |
| [Datadog.Trace.ClrProfiler.AutoInstrumentation.Wcf.TaskMethodInvokerIntegration](https://github.com/DataDog/dd-trace-dotnet/tree/d6d45a5d3fc9e8c60ae9b1be3e25dc403acf3290/tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Wcf/TaskMethodInvokerIntegration.cs) | `-12%` :no_entry: | `-8%` :no_entry:   | `2` :no_entry: |
| [Datadog.Trace.AsyncLocalScopeManager](https://github.com/DataDog/dd-trace-dotnet/tree/d6d45a5d3fc9e8c60ae9b1be3e25dc403acf3290/tracer/src/Datadog.Trace/AsyncLocalScopeManager.cs) | `-11%` :no_entry: | `0%` :heavy_check_mark:   | `8` :no_entry: |
| [Datadog.Trace.Util.ThrowHelper](https://github.com/DataDog/dd-trace-dotnet/tree/d6d45a5d3fc9e8c60ae9b1be3e25dc403acf3290/tracer/src/Datadog.Trace/Util/ThrowHelper.cs) | `-9%` :no_entry: | `0%` :heavy_check_mark:   | `0` :heavy_check_mark: |
| [Datadog.Trace.ClrProfiler.AutoInstrumentation.Testing.NUnit.NUnitCompositeWorkItemSkipChildrenIntegration](https://github.com/DataDog/dd-trace-dotnet/tree/d6d45a5d3fc9e8c60ae9b1be3e25dc403acf3290/tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Testing/NUnit/NUnitCompositeWorkItemSkipChildrenIntegration.cs) | `-8%` :no_entry: | `-8%` :no_entry:   | `0` :heavy_check_mark: |


The following classes were added in #2183:

| File    | Line coverage | Branch coverage | Complexity |
|:--------|:--------------------:|:----------:|:--------:|
| [Datadog.Trace.ClrProfiler.AutoInstrumentation.MongoDb.BsonDocumentProxy](https://github.com/DataDog/dd-trace-dotnet/pull/2183/files) | `0%` | `100%` | `1` |


11 classes were removed from Datadog.Trace in #2183


View the full reports for further details:

* [HTML report for master](https://dev.azure.com/datadoghq/dd-trace-dotnet/_build/results?buildId=84518&view=codecoverage-tab) | [Source file](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/84518/artifacts?artifactName=Code%20Coverage%20Report_84518&api-version=6.0&%24format=zip)  
* [HTML report for this PR #2183](https://dev.azure.com/datadoghq/dd-trace-dotnet/_build/results?buildId=84562&view=codecoverage-tab) | [Source file](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/84562/artifacts?artifactName=Code%20Coverage%20Report_84562&api-version=6.0&%24format=zip)

**Comment by andrewlock** [comment]
## Code Coverage Report :bar_chart:

:heavy_check_mark: Merging #2183 into [master](https://github.com/DataDog/dd-trace-dotnet/tree/b17a9fd4f332efd23af6b4974350f86a54297a22) will will **increase** line coverage by `1%`
:heavy_check_mark: Merging #2183 into [master](https://github.com/DataDog/dd-trace-dotnet/tree/b17a9fd4f332efd23af6b4974350f86a54297a22) will will **increase** branch coverage by `1%`
:heavy_check_mark: Merging #2183 into [master](https://github.com/DataDog/dd-trace-dotnet/tree/b17a9fd4f332efd23af6b4974350f86a54297a22) will will **decrease** complexity by `358`


|           | [master](https://github.com/DataDog/dd-trace-dotnet/tree/b17a9fd4f332efd23af6b4974350f86a54297a22) | #2183       | Change   | 
|:----------|:-----------:|:-----------:|:--------:|
| Lines     | `9334` / `12808` | `8955` / `12197` |          |
| Lines %   | `73%`      | `73%`      |  `1%` :heavy_check_mark:  |
| Branches  | `4246` / `6260` | `4075` / `5917` |          |
| Branches %| `68%`      | `69%`      |  `1%` :heavy_check_mark:  |
| Complexity|   `7456`      | `7098`        |  `-358`  :heavy_check_mark:    |

View the full report for further details:

* [HTML report for master](https://dev.azure.com/datadoghq/dd-trace-dotnet/_build/results?buildId=84518&view=codecoverage-tab) | [Source file](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/84518/artifacts?artifactName=Code%20Coverage%20Report_84518&api-version=6.0&%24format=zip)  
* [HTML report for this PR #2183](https://dev.azure.com/datadoghq/dd-trace-dotnet/_build/results?buildId=84570&view=codecoverage-tab) | [Source file](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/84570/artifacts?artifactName=Code%20Coverage%20Report_84570&api-version=6.0&%24format=zip)

### Datadog.Trace Breakdown :heavy_check_mark:

|        | [master](https://github.com/DataDog/dd-trace-dotnet/tree/b17a9fd4f332efd23af6b4974350f86a54297a22) | #2183       | Change   | 
|:-------|:-----------:|:-----------:|:--------:|
| Lines %| `73%`      | `73%`       |  `1%` :heavy_check_mark:  |
| Branches %| `68%`      | `69%`       |  `1%` :heavy_check_mark:  |
| Complexity| `7456`      | `7098`       |  `-358` :heavy_check_mark:  |

The following classes have significant coverage changes.

| File    | Line coverage change | Branch coverage change | Complexity change |
|:--------|:--------------------:|:----------:|:--------:|
| [Datadog.Trace.ClrProfiler.AutoInstrumentation.Couchbase.IIOServiceExecuteIntegrationBis](https://github.com/DataDog/dd-trace-dotnet/tree/998d3efeaf27e106a8e2cb0ade5c0281143ce588/tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Couchbase/IIOServiceExecuteIntegrationBis.cs) | `-100%` :no_entry: | `0%` :heavy_check_mark:   | `0` :heavy_check_mark: |
| [Datadog.Trace.ClrProfiler.AutoInstrumentation.Testing.MsTestV2.TestMethodRunnerExecuteIntegration](https://github.com/DataDog/dd-trace-dotnet/tree/998d3efeaf27e106a8e2cb0ade5c0281143ce588/tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Testing/MsTestV2/TestMethodRunnerExecuteIntegration.cs) | `-33%` :no_entry: | `-19%` :no_entry:   | `0` :heavy_check_mark: |
| [Datadog.Trace.ClrProfiler.AutoInstrumentation.Wcf.SyncMethodInvokerIntegration](https://github.com/DataDog/dd-trace-dotnet/tree/998d3efeaf27e106a8e2cb0ade5c0281143ce588/tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Wcf/SyncMethodInvokerIntegration.cs) | `-12%` :no_entry: | `-8%` :no_entry:   | `2` :no_entry: |
| [Datadog.Trace.ClrProfiler.AutoInstrumentation.Wcf.TaskMethodInvokerIntegration](https://github.com/DataDog/dd-trace-dotnet/tree/998d3efeaf27e106a8e2cb0ade5c0281143ce588/tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Wcf/TaskMethodInvokerIntegration.cs) | `-12%` :no_entry: | `-8%` :no_entry:   | `2` :no_entry: |
| [Datadog.Trace.AsyncLocalScopeManager](https://github.com/DataDog/dd-trace-dotnet/tree/998d3efeaf27e106a8e2cb0ade5c0281143ce588/tracer/src/Datadog.Trace/AsyncLocalScopeManager.cs) | `-11%` :no_entry: | `0%` :heavy_check_mark:   | `8` :no_entry: |
| [Datadog.Trace.Util.ThrowHelper](https://github.com/DataDog/dd-trace-dotnet/tree/998d3efeaf27e106a8e2cb0ade5c0281143ce588/tracer/src/Datadog.Trace/Util/ThrowHelper.cs) | `-9%` :no_entry: | `0%` :heavy_check_mark:   | `0` :heavy_check_mark: |
| [Datadog.Trace.ClrProfiler.AutoInstrumentation.Testing.NUnit.NUnitCompositeWorkItemSkipChildrenIntegration](https://github.com/DataDog/dd-trace-dotnet/tree/998d3efeaf27e106a8e2cb0ade5c0281143ce588/tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Testing/NUnit/NUnitCompositeWorkItemSkipChildrenIntegration.cs) | `-8%` :no_entry: | `-8%` :no_entry:   | `0` :heavy_check_mark: |


The following classes were added in #2183:

| File    | Line coverage | Branch coverage | Complexity |
|:--------|:--------------------:|:----------:|:--------:|
| [Datadog.Trace.ClrProfiler.AutoInstrumentation.MongoDb.BsonDocumentProxy](https://github.com/DataDog/dd-trace-dotnet/pull/2183/files) | `0%` | `100%` | `1` |


12 classes were removed from Datadog.Trace in #2183


View the full reports for further details:

* [HTML report for master](https://dev.azure.com/datadoghq/dd-trace-dotnet/_build/results?buildId=84518&view=codecoverage-tab) | [Source file](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/84518/artifacts?artifactName=Code%20Coverage%20Report_84518&api-version=6.0&%24format=zip)  
* [HTML report for this PR #2183](https://dev.azure.com/datadoghq/dd-trace-dotnet/_build/results?buildId=84570&view=codecoverage-tab) | [Source file](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/84570/artifacts?artifactName=Code%20Coverage%20Report_84570&api-version=6.0&%24format=zip)

File: tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/MongoDb/IConnection.cs
Comment (andrewlock) [reply tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/MongoDb/IConnection.cs]: @tonyredondo I don't quite follow either. The two options AFAIK are:

1. Use interfaces as above, and use constraint in the integration. If I understand you correctly, you're saying that the actual type created will be a struct that implements the proxy interface, so we get the "best of both worlds" in that case - it's a struct, but we can also call methods (which we can't do with a `[DuckCopy]` struct
2. As this type doesn't use any methods, we can a `[DuckCopy]` struct, but we can't use it in a constraint, so would need to do the duck typing _inside_ the integration, instead of as part of the CallTarget functionality. 

Presumably, in this case, the performance is similar (I assume we don't `[DuckCopy]` if it's a "constraint struct", does that matter?), and in some cases (where we need to call methods) 1. will be the only option. In which case I can't really see a case where 2. would be preferable?

>the limitation is that we cannot compare with null unless we also apply the IDuckType interface as a constraint

Is there a downside to applying the `IDuckType` constraint, as the type always implements it, correct?

Comment (tonyredondo) [reply tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/MongoDb/IConnection.cs]: I created this PR, with the best practices and benchmarks for all options:

https://github.com/DataDog/dd-trace-dotnet/pull/2223

Comment (zacharycmontoya) [reply tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/MongoDb/IConnection.cs]: Looks like the best way to handle this is to keep the interface with generic constraint, so I'll keep this type as-is

**Pull request merged**
